ops/hookcmds.py

```diff
@@ -0,0 +1,995 @@
+    id: int | None = None,
+    *,
+    app: bool = False,
+    file: pathlib.Path | None = None,
+):
+    """Set relation settings.
+
+    Note that ``id`` can only be ``None`` if the current hook is a relation
+    event, in which case Juju will use the ID of the relation that triggered the
+    event.
+    """
+    args = ['relation-set']
+    if app:
+        args.append('--app')
+    if id is not None:
+        args.extend(['-r', str(id)])
+    if file is not None:
+        data = dict(data)
+        with open(file) as f:
+            data.update(yaml.safe_load(f))
+    args.extend(['--file', '-'])
+    content = yaml.safe_dump(data)
+    _run(*args, input_stream=content)
+
+
+def resource_get(name: str) -> pathlib.Path:
+    """Get the path to the locally cached resource file."""
+    out = _run('resource-get', name, return_output=True)
+    return pathlib.Path(out.strip())
+
+
+def secret_add(
+    content: dict[str, str],
+    *,
+    label: str | None = None,
+    description: str | None = None,
+    expire: datetime.datetime | None = None,
+    rotate: SecretRotate | None = None,
+    owner: str | None = None,
+) -> str:
+    """Add a new secret."""
+    args: list[str] = []
+    if label is not None:
+        args.extend(['--label', label])
+    if description is not None:
+        args.extend(['--description', description])
+    if expire is not None:
+        args.extend(['--expire', expire.isoformat()])
+    if rotate is not None:
+        args.extend(['--rotate', rotate.value])
+    if owner is not None:
+        args.extend(['--owner', owner])
+    with tempfile.TemporaryDirectory() as tmp:
+        for k, v in content.items():
+            with open(f'{tmp}/{k}', mode='w', encoding='utf-8') as f:
+                f.write(v)
+            args.append(f'{k}#file={tmp}/{k}')
+        result = _run('secret-add', *args, return_output=True)
+    return result.strip()
+
+
+@overload
+def secret_get(
+    *,
+    id: str,
+    refresh: bool = False,
+    peek: bool = False,
+) -> dict[str, str]: ...
+@overload
+def secret_get(
+    *,
+    label: str,
+    refresh: bool = False,
+    peek: bool = False,
+) -> dict[str, str]: ...
+def secret_get(
+    *,
+    id: str | None = None,
+    label: str | None = None,
+    refresh: bool = False,
+    peek: bool = False,
+) -> dict[str, str]:
+    """Get the content of a secret."""
+    args: list[str] = []
+    if id is not None:
+        args.append(id)
+    if label is not None:
+        args.extend(['--label', label])
+    if refresh:
+        args.append('--refresh')
+    if peek:
+        args.append('--peek')
+    result = _run('secret-get', *args, return_output=True, use_json=True)
+    return cast('dict[str, str]', result)
+
+
+def secret_grant(id: str, relation_id: int, *, unit: str | None = None):
+    """Grant access to a secret."""
+    args = ['secret-grant', '--relation', str(relation_id)]
+    if unit is not None:
+        args += ['--unit', str(unit)]
```

Comment:
Nit: while we're in here, let's use `list.extend` consistently (or `+=` consistently, but I think the majority is extend).